### PR TITLE
add error handling for config.rs && add more unit tests  && ip_match

### PR DESCRIPTION
**Pull request by GopherJ**

### CODE REVIEW

**Comment by GopherJ** [comment]
hello @hsluoyz  @xcaptain could you review the changes?

**Comment by hsluoyz** [comment]
Hi @GopherJ , can you remove the merge commits like:

image

You can rebase on origin/master HEAD first, then apply your commits via cherry-pick.

**Comment by GopherJ** [comment]
hello @hsluoyz , could you check if it's fine now? I didn't use `cherry-pick` before so I'm not sure if everything is fine, let me know if something need to be changed~ good night!

**Comment by hsluoyz** [comment]
These two commits don't seem to be right:

image

**Comment by GopherJ** [comment]
Ok I’ll remove these two tomorrow

File: src/config.rs [hunk 1]
```diff
@@ -127,28 +138,26 @@ impl Config {
         }
     }
 
-    pub fn get(&self, key: &str) -> String {
+    pub fn get(&self, key: &str) -> Option<&str> {
```
Comment (xcaptain) [inline src/config.rs]: I would prefer panic if the config is not set correctly, Option or Result should be used in runtime error.

**Pull request closed without merging**